src/hotspot/share/runtime/thread.cpp

```diff
@@ -1743,18 +1766,20 @@ void JavaThread::send_thread_stop(oop java_throwable)  {
 //   - Target thread will not enter any new monitors.
 //
 bool JavaThread::java_suspend() {
-  ThreadsListHandle tlh;
-  if (!tlh.includes(this)) {
-    log_trace(thread, suspend)("JavaThread:" INTPTR_FORMAT " not on ThreadsList, no suspension", p2i(this));
+  guarantee(Thread::is_JavaThread_protected_by_my_ThreadsList(this),
+            "missing ThreadsListHandle in calling context.");
+  if (is_exiting()) {
```

Comment:
Ummmm... The purpose of the new `is_exiting()` check and the baseline's
`ThreadsListHandle::includes()` check is to avoid making this call:
```
  return this->handshake_state()->suspend();
```
The call we are avoiding is the one that makes the synchronous
SuspendThreadHandshake request (for threads other than self)
so by detecting the exiting thread early, we are avoiding entering
the handshake machinery.

Reply:
But why do we care about that? The low-level mechanism handles exiting threads so there is no need for the higher-level code to do that. The current logic gives the appearance that we must filter out exiting threads at this level because the lower-level code does not handle them.

Reply:
Okay you've convinced me. I'm merging my latest patch with master
and then I'll retest with the is_exiting() check removed.